amm-executor/src/Spectrum/Executor/Backlog/Service.hs

```diff
@@ -1,60 +1,174 @@
   pendingPQ   <- newIORef PQ.empty  -- ordered by weight; new orders
   suspendedPQ <- newIORef PQ.empty  -- ordered by weight; failed orders, waiting for retry (retries are performed with some constant probability, e.g. 5%) 
   toRevisitQ  <- newIORef Seq.empty -- regular queue; successully submitted orders. Left orders should be re-executed in X minutes. Normally successfully confirmed orders are eliminated from this queue.
+  _           <- liftK $ recover store config pendingPQ
+  pure $ attachLogging logging BacklogService
+    { put = \(PendingOrder order timestamp) -> do
+        put $ BacklogOrder timestamp order
+        modifyIORef pendingPQ (PQ.insert $ mkWeightedOrderWithTimestamp order timestamp)
+    , suspend = \(SuspendedOrder order timestamp) -> do
+        modifyIORef suspendedPQ (PQ.insert $ mkWeightedOrderWithTimestamp order timestamp)
+        exists $ BacklogOrder timestamp order
+    , checkLater = \(InProgressOrder order timestamp) -> do
+        modifyIORef toRevisitQ (mkWeightedOrderWithTimestamp order timestamp Seq.<|)
+        exists $ BacklogOrder timestamp order
+    , tryAcquire = do
+        _ <- refreshQueues config store pendingPQ toRevisitQ
+        maxWeightedOrderM <- getMaxWeightedOrder' config store pendingPQ suspendedPQ
+        case maxWeightedOrderM of
+          Just (WeightedOrderWithTimestamp oId _ _) -> get oId <&> fmap backlogOrder
+          Nothing -> pure Nothing
+    , drop = \oId -> do
+        _ <- modifyIORef pendingPQ   (PQ.filter (\(WeightedOrderWithTimestamp orderId _ _) -> orderId /= oId))
+        _ <- modifyIORef suspendedPQ (PQ.filter (\(WeightedOrderWithTimestamp orderId _ _) -> orderId /= oId))
+        _ <- modifyIORef toRevisitQ  (Seq.filter (\(WeightedOrderWithTimestamp orderId _ _) -> orderId /= oId))
+        drop oId
+    }
 
-  (_, db) <- Rocks.openBracket storePath
-              Rocks.defaultOptions
-                { Rocks.createIfMissing = createIfMissing
-                }
+refreshQueues
+  :: (MonadIO m)
+  => BacklogServiceConfig
+  -> BacklogStore m
+  -> IORef (PQ.MaxQueue WeightedOrderWithTimestamp)
+  -> IORef (Seq.Seq WeightedOrderWithTimestamp)
+  -> m ()
+refreshQueues BacklogServiceConfig{..} BacklogStore{..} pendingQueueRef toRevisitSeqRef = do
+  currentTime <- getCurrentTime
+  revisited2process <-
+    atomicModifyIORef'
+      toRevisitSeqRef
+      (Seq.spanl (\(WeightedOrderWithTimestamp _ _ oTime) -> diffUTCTime currentTime oTime < orderExecTime))
+  mapM_ (\order@(WeightedOrderWithTimestamp oId _ oTime) ->
+          if diffUTCTime currentTime oTime < orderLifetime
+            then modifyIORef' pendingQueueRef (PQ.insert order)
+            else drop oId
+       ) revisited2process
+
+getMaxWeightedOrder'
+  :: forall m. (MonadIO m)
+  => BacklogServiceConfig
+  -> BacklogStore m
+  -> IORef (PQ.MaxQueue WeightedOrderWithTimestamp)
+  -> IORef (PQ.MaxQueue WeightedOrderWithTimestamp)
```

Comment:
The idea was to keep only (Weight, OrderId) in queues and orders themselves in persistent storage. Then you have only one mandatory source of truth.